### DISCUSSION TITLE
'Type parameter "T@Bar" is covariant, but "T@Bar" is not a subtype of "Jee"'. But `T: Jee`...?

### DISCUSSION THREAD

**Discussion by qexat** [microsoft/pyright#8993] [Q&A] [answered]
The following snippet:
```py
from __future__ import annotations

import dataclasses


class Foo:
    def bark(self) -> Jee:
        return self


@dataclasses.dataclass(slots=True, frozen=True)
class Bar[T: Jee]:  # T is clearly a Jee.
    meow: T

    def bark(self) -> Jee:
        return self  # error


type Jee = Foo | Bar[Jee]
```

causes Pyright to report the following:
```
  repro.py:16:16 - error: Type "Self@Bar[T@Bar]" is not assignable to return type "Jee"
    Type "Self@Bar[T@Bar]" is not assignable to type "Jee"
      "Bar[T@Bar]*" is not assignable to "Foo"
      "Bar[T@Bar]*" is not assignable to "Bar[Jee]"
        Type parameter "T@Bar" is covariant, but "T@Bar" is not a subtype of "Jee"
          Type "T@Bar" is not assignable to type "Jee"
            Type "T@Bar" is not assignable to type "Foo"
            Type "T@Bar" is not assignable to type "Bar[Jee]" (reportReturnType)
1 error, 0 warnings, 0 informations
```

I am confused. Am I missing something?

**Comment by erictraut** [comment, marked as answer]
Pyright's behavior here is correct. (Not surprisingly, mypy generates the same error in this case.)

You haven't provided a type annotation for the `self` parameter in the `bark` method, so its type is implicitly `typing.Self`. If your intent is for it to be `Jee`, you can annotate it as such, but that means this method won't be callable for certain specializations of `Bar`. If you want `bark` to work with all parameterizations of `Bar`, then you'd need to change the return type of `bark` to `Self`.

**Reply by qexat** [reply]
> Pyright's behavior here is correct. (Not surprisingly, mypy generates the same error in this case.)
> 
> You haven't provided a type annotation for the `self` parameter in the `bark` method, so its type is implicitly `typing.Self`. If your intent is for it to be `Jee`, you can annotate it as such, but that means this method won't be callable for certain specializations of `Bar`. If you want `bark` to work with all parameterizations of `Bar`, then you'd need to change the return type of `bark` to `Self`.

Oh oops I guess I should have slept more :') Thank you for the answer!
